### PR TITLE
Add step to eval code on instances

### LINKED ISSUE #287: [2pt] Add a step to run provieded lua-script on cluster instancess

### PR DESCRIPTION
**Pull request by dokshina**
Closes #287

### CODE REVIEW

File: defaults/main.yml [hunk 1]
```diff
@@ -126,3 +126,9 @@ control_instance: null
 temporary_files: []
 needs_restart: null
 not_expelled_instance: null
+
+# Eval params
+
+cartridge_eval_body: null
+cartridge_eval_args: []
+cartridge_eval_file: null
```
Comment (hackallcode) [inline defaults/main.yml]:
```suggestion
cartridge_eval_file: null
cartridge_eval_body: null
cartridge_eval_args: []
```

File: doc/eval.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+        return unpack({
+            string.format("Hi, %s!", name),
+            string.format("You are %s years old and live in %s.", bio.age, bio.city),
+        })
+
+    cartridge_eval_args:
+      - "Elizabeth"
+      - age: 24
+        city: Moscow
```
Comment (hackallcode) [inline doc/eval.md]:
```suggestion
        city: 'Moscow'
```

File: molecule/eval/hosts.yml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  hosts:
+    storage-1-leader:
+      config:
+        advertise_uri: 'vm1:3102'
+        http_port: 8102
+
+    storage-1-replica:
+      config:
+        advertise_uri: 'vm1:3103'
+        http_port: 8103
```
Comment (hackallcode) [inline molecule/eval/hosts.yml]:
```suggestion
    storage-1-leader:
      config:
        advertise_uri: 'vm1:3101'
        http_port: 8101

    storage-1-replica:
      config:
        advertise_uri: 'vm1:3102'
        http_port: 8102
```

File: tasks/set_instance_facts.yml [hunk 4]
```diff
@@ -96,3 +96,9 @@
       temporary_files: '{{ temporary_files }}'
       needs_restart: '{{ needs_restart }}'
       not_expelled_instance: '{{ not_expelled_instance }}'
+
+      # Eval params
+
+      cartridge_eval_body: '{{ cartridge_eval_body }}'
+      cartridge_eval_args: '{{ cartridge_eval_args }}'
+      cartridge_eval_file: '{{ cartridge_eval_file }}'
```
Comment (hackallcode) [inline tasks/set_instance_facts.yml]:
```suggestion
      cartridge_eval_file: '{{ cartridge_eval_file }}'
      cartridge_eval_body: '{{ cartridge_eval_body }}'
      cartridge_eval_args: '{{ cartridge_eval_args }}'
```

File: tasks/step_cleanup.yml [hunk 5]
```diff
@@ -6,4 +6,4 @@
     - cartridge-replicasets
     - cartridge-config
 
-- import_tasks: 'steps/cleanup.yml'
+- import_tasks: 'steps/eval.yml'
```
Comment (hackallcode) [inline tasks/step_cleanup.yml]: Typo?

File: tasks/steps/eval.yml [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  cartridge_eval:
+    console_sock: '{{ instance_info.console_sock }}'
+    body: '{{ cartridge_eval_body }}'
+    args: '{{ cartridge_eval_args }}'
+  register: eval_res_register
+  when: cartridge_eval_body is not none
+
+- name: 'Set "eval_res" fact'
+  set_fact:
+    eval_res: '{{ eval_res_register.fact | default(none) }}'
```
Comment (hackallcode) [inline tasks/steps/eval.yml]:
```suggestion
    eval_res: '{{ eval_res_register.fact | default(none) }}'
  when: cartridge_eval_body is not none
```

File: tasks/steps/eval.yml [hunk 7]
```diff
@@ -0,0 +1,18 @@
+---
+
+- name: 'Read code from file'
+  set_fact:
+    cartridge_eval_body: '{{ lookup("file", cartridge_eval_file) }}'
```
Comment (hackallcode) [inline tasks/steps/eval.yml]: It won't work in the following case:
```yaml
- name: Eval function from file
  hosts: cluster
  roles:
    - ansible-cartridge
  vars:
    cartridge_scenario:
      - eval
    cartridge_eval_file: "test.yml"

- name: Eval function from string
  hosts: cluster
  roles:
    - ansible-cartridge
  vars:
    cartridge_scenario:
      - eval
    cartridge_eval_body: "return 123"  # !!! will be run "test.yml"
```

File: molecule/eval/converge.yml [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        success_msg: 'Received result is OK'
+        that: >-
+          eval_res == [
+              "This snippet was evaled from string",
+              "Hello, my name is " + instance_info.instance_id + "!",
+              "I am 43 seconds old",
+              "--- {'cute': 'table'}\n...\n",
+          ]
+
+## EVAL FROM FILE
```
Comment (hackallcode) [inline molecule/eval/converge.yml]: Let's swap the tests, so it will be tested that the body can be rewritten via vars after file eval use!

**Review by hackallcode — CHANGES_REQUESTED** [review]

File: molecule/eval/converge.yml (hunk 8 above)
Comment (dokshina) [reply molecule/eval/converge.yml]: As I see, you know all about such cases =)

Thanks!

File: tasks/steps/eval.yml [hunk 9]
```diff
@@ -2,17 +2,22 @@
 
 - name: 'Read code from file'
   set_fact:
-    cartridge_eval_body: '{{ lookup("file", cartridge_eval_file) }}'
-  when: cartridge_eval_file is not none and cartridge_eval_body is none
+    eval_body: >-
+      {{
+        cartridge_eval_body
+        if (cartridge_eval_body is not none)
+        else lookup("file", cartridge_eval_file)
```
Comment (hackallcode) [inline tasks/steps/eval.yml]: If I have cartridge_eval_body == null and cartridge_eval_file == null, then it will look up null file

File: doc/eval.md [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+  become_user: root
+  gather_facts: false
+  vars:
+    cartridge_scenario:
+      - eval
+    cartridge_eval_file: my_snippet.lua
+    cartridge_eval_args:
+      - "Elizabeth"
+      - age: 24
+        city: Moscow
```
Comment (hackallcode) [inline doc/eval.md]:
```suggestion
        city: 'Moscow'
```

File: molecule/eval/converge.yml [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+  become: true
+  become_user: root
+  gather_facts: false
+  vars:
+    cartridge_app_name: 'myapp'
+    cartridge_scenario:
+      - eval
+    cartridge_eval_file: some_script.lua
+    cartridge_eval_args:
+      - Hello, my name is %s!
```
Comment (hackallcode) [inline molecule/eval/converge.yml]: And the same comment below:
```suggestion
      - 'Hello, my name is %s!'
```

File: molecule/eval/converge.yml [hunk 12]
```diff
@@ -1,5 +1,19 @@
 ---
 
+# EMPTY EVAL
```
Comment (hackallcode) [inline molecule/eval/converge.yml:3]: Let's move it to the end of file

File: tasks/steps/eval.yml [hunk 13]
```diff
@@ -7,6 +7,8 @@
         cartridge_eval_body
         if (cartridge_eval_body is not none)
         else lookup("file", cartridge_eval_file)
+          if cartridge_eval_file is not none
```
Comment (hackallcode) [inline tasks/steps/eval.yml]: Let's use parentheses for visibility of separation

File: molecule/eval/converge.yml [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+      assert:
+        msg: 'Eval should be skipped'
+        success_msg: 'Eval was skipped'
+        that: eval_res_register.skipped | default(none)
+
+    - name: Check that eval result is empty
+      assert:
+        msg: 'Eval result should be empty'
+        success_msg: 'Eval result is empty'
+        that: eval_res == ""
```
Comment (hackallcode) [inline molecule/eval/converge.yml]: > Be careful, **`eval_res` is always a list**

But here `eval_res` is a string... Maybe it's ok, but strange

Comment (dokshina) [reply molecule/eval/converge.yml]: Look

```
  set_fact:
    eval_res: '{{ eval_res_register.fact | default(none) }}'
```

**Review by hackallcode — DISMISSED** [review]

File: .github/helpers/count_molecule_matrix.py [hunk 15]
```diff
@@ -31,7 +31,12 @@ def main(event_name, repo_owner, review_state, ref):
     ee_matrix = []
 
     if event_name == 'push' or event_name == 'pull_request' and repo_owner != 'tarantool':
-        ce_matrix.append(get_ce_params())
+        # ce_matrix.append(get_ce_params())
+
+        # remove before merge
+        ce_matrix.append(get_ce_params(molecule_scenario='eval', ansible_version='2.8.0'))
+        ce_matrix.append(get_ce_params(molecule_scenario='eval', ansible_version='2.9.0'))
+        ce_matrix.append(get_ce_params(molecule_scenario='eval', ansible_version='2.10.0'))
```
Comment (hackallcode) [inline .github/helpers/count_molecule_matrix.py]: Don't forget undo this changes

**Review by hackallcode — DISMISSED** [review]

**Review by hackallcode — APPROVED** [review]
